src/xcode/ENA/ENA/Source/Services/Risk/Provider/Helper/CachedAppConfiguration.swift

```diff
@@ -20,48 +20,87 @@
+
+		// edge case: if no app config is cached, omit a potentially existing ETag to force fetch a new configuration
+		let etag = store.appConfig == nil ? nil : store.lastAppConfigETag
+
+		// check for updated or fetch initial app configuration
+		fetchConfig(with: etag)
 	}
 
-	// MARK: Properties
-	private let client: Client
-	private var cache: Cache?
+	private func fetchConfig(with etag: String?, completion: Completion? = nil) {
+		client.fetchAppConfiguration(etag: etag) { [weak self] result in
+			switch result {
+			case .success(let response):
+				self?.store.lastAppConfigETag = response.eTag
+				self?.store.appConfig = response.config
+				completion?(.success(response.config))
+			case .failure(let error):
+				switch error {
+				case CachedAppConfiguration.CacheError.notModified where self?.store.appConfig != nil:
+					log(message: "config not modified")
+					// server is not modified and we have a cached config
+					guard let config = self?.store.appConfig else {
+						fatalError("App configuration cache broken!") // in `where` we trust
+					}
+					completion?(.success(config))
+
+					// keep track of last successful fetch
+					self?.store.lastAppConfigFetch = Date()
+				default:
+					completion?(.failure(error))
+				}
+			}
+		}
+	}
 }
 
 extension CachedAppConfiguration: AppConfigurationProviding {
-	func appConfiguration(completion: @escaping Completion) {
-		guard let cache = cache else {
-			actuallyDownloadAppConfiguration(completion: completion)
-			return
-		}
 
-		let calendar = Calendar.current
-		let deltaInMinutes = abs(calendar.dateComponents([.minute], from: cache.date, to: Date()).minute ?? .max)
-		if deltaInMinutes < 5 {
-			completion(cache.value)
-			return
+	fileprivate static let timestampKey = "LastAppConfigFetch"
+
+	func appConfiguration(forceFetch: Bool = false, completion: @escaping Completion) {
+		let force = shouldFetch() || forceFetch
+
+		if let cachedVersion = store.appConfig, !force {
+			// use the cached version
+			completion(.success(cachedVersion))
+		} else {
+			// fetch a new one
+			fetchConfig(with: store.lastAppConfigETag, completion: completion)
 		}
-		actuallyDownloadAppConfiguration(completion: completion)
 	}
 
-	private func actuallyDownloadAppConfiguration(completion: @escaping Completion) {
-		client.appConfiguration { [weak self] appConfiguration in
-			guard let appConfiguration = appConfiguration else {
-				self?.cache = nil
-				completion(nil)
-				return
-			}
-			self?.cache = .init(date: Date(), value: appConfiguration)
-			completion(appConfiguration)
-		}
+	func appConfiguration(completion: @escaping Completion) {
+		self.appConfiguration(forceFetch: false, completion: completion)
 	}
-}
 
-private extension CachedAppConfiguration {
-	struct Cache {
-		let date: Date
-		let value: SAP_ApplicationConfiguration
+	/// Simple helper to simulate Cache-Control
+	/// - Note: This 300s value is because of current handicaps with the HTTPClient
```

Comment:
Could you please explain, if possible, which handicaps?

Reply:
Bad architecture. I would have to refactor even more within the HTTPClient to pass HTTP responses including header fields.

Reply:
Could you just mention the problem in the comment? And then its fine for me.

Reply:
Ok?

> /// Simple helper to simulate Cache-Control
	/// - Note: This 300 second value is because of current handicaps with the HTTPClient architecture
	///   which does not easily return response headers. This requires further refactoring of `URLSession+Convenience.swift`.
	/// - Returns: `true` is a network call should be done; `false` if cache should be used

Reply:
Thanks!